framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py

```diff
@@ -0,0 +1,218 @@
+      yExpRef = np.mean(yExp)
+      yAppRef = np.mean(yApp)
+      yExpStd = (yExp-yExpRef)/yExpRef
+      yAppStd = (yApp-yAppRef)/yAppRef
+
+      # Kernel Desnity Estimation
+      m1 = yExpStd[:]
+      m2 = yAppStd[:]
+      xmin = m1.min()
+      xmax = m1.max()
+      ymin = m2.min()
+      ymax = m2.max()
+
+      binKDE = 200j
+      X, Y = np.mgrid[xmin:xmax:binKDE, ymin:ymax:binKDE]
+      psts = np.vstack([X.ravel(), Y.ravel()])
+      vals = np.vstack([m1, m2])
+      # kernel
+      knl = stats.gaussian_kde(vals)
+      Z = np.reshape(knl(psts).T, X.shape)
+
+      # Virtual Measurement
+      msrMean = 0.0
+      msrStd = 0.01*np.std(yAppStd)
+      msrNumSmpl = 1000
+      randomUtils.randomSeed(0)
+      yMsr = msrStd * randomUtils.randomNormal(size=msrNumSmpl) + msrMean
+      binMsr = msrNumSmpl//20
+      yMsrPdf, yMsrBin = np.histogram(yMsr, binMsr, range=(xmin, xmax), density=True)
+
+      # yAppPred by integrating f(yexp, yapp)dyexp * f(yexp)
+      # on range [ymsr.min(), ymsr.max()]
+      yAppPredPdf = np.zeros(Y.shape[1])
+      intgrPdf = 0.0 # for normalization
+      for i in range(len(Y[0, :])):
+        yAppPi = 0.0
+        for j in range(len(yMsrBin)-1):
+          yAppPi += knl.evaluate([yMsrBin[j]+0.5*np.diff(yMsrBin)[j], Y[0, i]]) * np.diff(yMsrBin)[j] * yMsrPdf[j]
+        yAppPredPdf[i] = yAppPi
+        intgrPdf += yAppPredPdf[i]*(Y[0, 1]-Y[0, 0])
+
+      # normalized PDF of predicted application
+      yAppPredPdfNorm = yAppPredPdf/intgrPdf
+
+      # Calculate Expectation (average value) of predicted application
+      predMean = 0.0
+      for i in range(len(Y[0, :])):
+        predMean += Y[0, i]*yAppPredPdfNorm[i]*(Y[0, 1]-Y[0, 0])
+
+      # Calculate Variance of predicted application
+      predVar = 0.0
+      for i in range(len(Y[0, :])):
+        predVar += (Y[0, i]-predMean)**2.0 * yAppPredPdfNorm[i]*(Y[0, 1]-Y[0, 0])
+
+      # Standard Deviation
+      predStd = np.sqrt(predVar)
+      priStd = np.std(yAppStd)
+      stdReduct = 1.0-predStd/priStd
+
+      # Generate distribution by yAppPredPdfNorm
+      predBins = np.insert(Y[0, :]+0.5*(Y[0, 1]-Y[0, 0]), 0, Y[0, 0]-0.5*(Y[0, 1]-Y[0, 0]))
+      predHist = tuple((yAppPredPdfNorm, predBins))
+      predDist = stats.rv_histogram(predHist)
+      yAppPred = predDist.rvs(size=len(yAppStd), random_state=randomUtils.randomSeed(0, seedBoth=True))
+
+      # Form tuple data for metrics
+      priData = tuple((yAppStd.reshape(len(yAppStd),1), targData[1]))
+      postData = tuple((yAppPred.reshape(len(yAppStd),1), targData[1]))
+      for metric in self.metrics:
+        #name = "{}_{}_{}".format(feat.split("|")[-1], targ.split("|")[-1], metric.estimator.name)
+        name = "pri_post_{}".format(metric.estimator.name)
+        outputDict[name] = metric.evaluate((priData, postData), multiOutput='raw_values')
+
+    return outputDict
+
+  def _getDataFromDataDict(self, datasets, var, names=None):
```

Comment:
this function should be moved to ValidationBase class.

Reply:
Agree.

Reply:
What should I do here?